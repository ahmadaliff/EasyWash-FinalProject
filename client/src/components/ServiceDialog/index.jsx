import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { FormattedMessage, FormattedNumber } from 'react-intl';

import { Add, AddShoppingCart, Close, LibraryBooksOutlined, Remove } from '@mui/icons-material';
import { Button, Dialog, DialogActions, DialogContent, DialogTitle, IconButton } from '@mui/material';

import { actionAddToCart } from '@pages/Cart/action';

import classes from '@components/ServiceDialog/style.module.scss';
import { useNavigate } from 'react-router-dom';
import { useState } from 'react';
import { actionSetOrder } from '@pages/Order/action';

const ServiceDialog = ({ handleClose, open, service }) => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const [quantity, setQuantity] = useState(1);

  const handleCreateOrder = () => {
    const cart = [{ serviceId: service?.id, quantity, serviceName: service?.name, servicePrice: service?.price }];
    dispatch(actionSetOrder(cart));
    navigate('/order');
  };

  return (
    <Dialog open={open} onClose={handleClose} fullWidth className={classes.dialog} data-testid="dialog-service">
      <DialogTitle fontSize="large">
        <b>
          <FormattedMessage id="app_service_datail" />
        </b>
      </DialogTitle>
      <IconButton
        aria-label="close"
        onClick={() => {
          handleClose();
          setQuantity(1);
        }}
        sx={{
          position: 'absolute',
          right: 8,
          top: 8,
        }}
      >
        <Close />
      </IconButton>
      <DialogContent>
        <div>
          <div className={classes.field}>
            <p className={classes.label}>
              <FormattedMessage id="app_service_name" />
            </p>
            <p className={classes.text}>{service?.name}</p>
          </div>
          <div className={classes.field}>
            <p className={classes.label}>
              <FormattedMessage id="app_service_price" />
            </p>
            <p className={classes.text}>
              <FormattedNumber
                value={service?.price}
                // eslint-disable-next-line react/style-prop-object
                style="currency"
                currency="IDR"
                minimumFractionDigits={0}
              />
            </p>
          </div>
        </div>
        <div className={classes.buttonQuantity}>
          <IconButton onClick={() => setQuantity(quantity - 1)} disabled={quantity - 1 === 0} data-testid="button-min">
            <Remove />
          </IconButton>
          {quantity}
          <IconButton onClick={() => setQuantity(quantity + 1)} data-testid="button-plus">
            <Add />
          </IconButton>
        </div>
      </DialogContent>
      <DialogActions className={classes.action}>
        <Button
          variant="contained"
          className={classes.dialogAction}
          startIcon={<AddShoppingCart />}
          onClick={() => dispatch(actionAddToCart(service?.id, quantity))}
          data-testid="button-add"
        >
          <FormattedMessage id="app_add_to_cart" />
        </Button>
        <Button
          variant="contained"
          className={classes.dialogAction}
          startIcon={<LibraryBooksOutlined />}
          onClick={handleCreateOrder}
          data-testid="button-create"
        >
          <FormattedMessage id="app_create_order" />
        </Button>
      </DialogActions>
    </Dialog>
  );
};

ServiceDialog.propTypes = {
  handleClose: PropTypes.func,
  open: PropTypes.bool,
  service: PropTypes.object,
};

export default ServiceDialog;
