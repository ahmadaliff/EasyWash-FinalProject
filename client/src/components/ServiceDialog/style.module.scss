@import '@styles/base';

.dialog {
  &:nth-child(5) {
    bottom: 0;
    &:nth-child(5) {
      position: absolute;
      width: 100%;
      top: unset;
      right: unset;
      bottom: 0;
      left: 0;
    }
  }
  .action {
    padding: 1rem;
    .dialogAction {
      background-color: var(--color-primary);
      border-radius: 10rem;
      padding: 0.25rem 1rem 0.25rem 1rem;
      text-transform: capitalize;
      font-size: small;
      @include apply-at-md-size {
        font-size: medium;
      }
      &:hover {
        background-color: var(--color-primary);
      }
    }
  }
  .buttonQuantity {
    @include directionX;
    gap: 0.5rem;
    align-items: center;
  }
}
.field {
  @include directionX;
  align-items: center;
  font-size: 0.75rem;
  font-weight: 500;
  .label {
    float: left;
    font-weight: 700;
    width: 10rem;
  }
  .text {
    width: 100%;
  }
}
