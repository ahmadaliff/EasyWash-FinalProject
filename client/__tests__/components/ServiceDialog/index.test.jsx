import { render, fireEvent } from '@utils/testHelper';
import { act } from 'react-dom/test-utils';
import ServiceDialog from '@components/ServiceDialog';

let wrapper;
const mockDispatch = jest.fn();
const mockNavigate = jest.fn();

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

let isOpen = true;

const mockProps = {
  open: isOpen,
  handleClose: () => (isOpen = false),
  service: {
    id: 1,
    name: 'service',
    price: 12000,
  },
};
beforeEach(() => {
  act(() => {
    wrapper = render(<ServiceDialog {...mockProps} />);
  });
  jest.clearAllMocks();
});

describe('ServiceDialog Component', () => {
  test('Correct render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('dialog-service')).toBeInTheDocument();
  });

  test('Calls handleClose when close button is clicked', () => {
    const { getByLabelText } = wrapper;
    act(() => {
      fireEvent.click(getByLabelText('close'));
    });
    expect(isOpen).toBe(false);
  });

  test('Updates quantity when +/- buttons are clicked', () => {
    const { getByTestId, getByText } = wrapper;
    act(() => {
      fireEvent.click(getByTestId('button-min'));
    });
    expect(getByText('1')).toBeInTheDocument();

    act(() => {
      fireEvent.click(getByTestId('button-plus'));
    });
    expect(getByText('2')).toBeInTheDocument();
  });

  test('Dispatches actionAddToCart when "Add to Cart" button is clicked', () => {
    const { getByTestId } = wrapper;
    act(() => {
      fireEvent.click(getByTestId('button-add'));
    });
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });

  test('Calls handleCreateOrder and navigates when "Create Order" button is clicked', () => {
    const { getByTestId } = wrapper;
    act(() => {
      fireEvent.click(getByTestId('button-create'));
    });
    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(mockNavigate).toHaveBeenCalledWith('/order');
  });

  test('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
