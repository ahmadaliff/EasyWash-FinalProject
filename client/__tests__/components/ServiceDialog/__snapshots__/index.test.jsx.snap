// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceDialog Component Matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDialog-root dialog MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
      data-testid="dialog-service"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby=":r5:"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-tlc64q-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-1xi8exm-MuiTypography-root-MuiDialogTitle-root"
            id=":r5:"
          >
            <b>
              Detail Layanan
            </b>
          </h2>
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-qsfli3-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
          >
            <div>
              <div
                class="field"
              >
                <p
                  class="label"
                >
                  Nama Layanan
                </p>
                <p
                  class="text"
                >
                  service
                </p>
              </div>
              <div
                class="field"
              >
                <p
                  class="label"
                >
                  Harga
                </p>
                <p
                  class="text"
                >
                  RpÂ 12.000
                </p>
              </div>
            </div>
            <div
              class="buttonQuantity"
            >
              <button
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                data-testid="button-min"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="RemoveIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13H5v-2h14v2z"
                  />
                </svg>
              </button>
              1
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                data-testid="button-plus"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing action css-knqc4i-MuiDialogActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium dialogAction css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="button-add"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddShoppingCartIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
                  />
                </svg>
              </span>
              Tambahkan ke keranjang
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium dialogAction css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="button-create"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="LibraryBooksOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M4 6H2v14c0 1.1.9 2 2 2h14v-2H4V6zm16-4H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H8V4h12v12zM10 9h8v2h-8zm0 3h4v2h-4zm0-6h8v2h-8z"
                  />
                </svg>
              </span>
              Buat Pesanan
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
